initrd/ociimage.go

```diff
@@ -0,0 +1,284 @@
+		OSChoice: "linux",
+	}
+
+	if initrd.opts.arch != "" {
+		if initrd.opts.arch == "x86_64" {
+			sysCtx.ArchitectureChoice = "amd64"
+		} else {
+			sysCtx.ArchitectureChoice = initrd.opts.arch
+		}
+	}
+
+	policy := &signature.Policy{
+		Default: []signature.PolicyRequirement{
+			signature.NewPRInsecureAcceptAnything(),
+		},
+	}
+
+	policyCtx, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate default policy context: %w", err)
+	}
+
+	defer policyCtx.Destroy()
+
+	img, err := initrd.ref.NewImage(ctx, sysCtx)
+	if err != nil {
+		return "", err
+	}
+
+	defer img.Close()
+
+	ociImage, err := img.OCIConfig(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	initrd.args = ociImage.Config.Entrypoint
+	initrd.args = append(initrd.args, ociImage.Config.Cmd...)
+	initrd.env = ociImage.Config.Env
+
+	if initrd.opts.output == "" {
+		fi, err := os.CreateTemp("", "")
+		if err != nil {
+			return "", err
+		}
+
+		initrd.opts.output = fi.Name()
+	}
+
+	// Create a temporary directory to output the image to
+	outputDir, err := os.MkdirTemp("", "")
+	if err != nil {
+		return "", fmt.Errorf("could not make temporary directory: %w", err)
+	}
+
+	defer os.RemoveAll(outputDir)
+
+	ociTarballFile := filepath.Join(outputDir, "oci.tar.gz")
+
+	dest, err := ociarchive.NewReference(ociTarballFile, "")
+	if err != nil {
+		return "", fmt.Errorf("invalid destination name %s: %v", dest, err)
+	}
+
+	opts := copy.Options{
+		ReportWriter:   log.G(ctx).Writer(),
+		DestinationCtx: sysCtx,
+		SourceCtx:      sysCtx,
+	}
+
+	if _, err = copy.Image(ctx, policyCtx, dest, initrd.ref, &opts); err != nil {
+		return "", fmt.Errorf("failed to copy image: %w", err)
+	}
+
+	image, err := stereoscope.GetImage(ctx, ociTarballFile)
+	if err != nil {
+		return "", fmt.Errorf("could not load image: %w", err)
+	}
+
+	f, err := os.OpenFile(initrd.opts.output, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0o644)
+	if err != nil {
+		return "", fmt.Errorf("could not open initramfs file: %w", err)
+	}
+
+	defer f.Close()
+
+	writer := cpio.NewWriter(f)
+	defer writer.Close()
+
+	if err := image.SquashedTree().Walk(func(path scfile.Path, f filenode.FileNode) error {
+		if f.Reference == nil {
+			log.G(ctx).
+				WithField("path", path).
+				Warn("skipping: no reference")
+			return nil
+		}
+
+		info, err := image.FileCatalog.Get(*f.Reference)
+		if err != nil {
+			return err
+		}
+
+		internal := fmt.Sprintf(".%s", path)
+
+		if f.FileType == scfile.TypeDirectory {
+			if err := writer.WriteHeader(&cpio.Header{
+				Name: internal,
+				Mode: cpio.FileMode(info.Mode().Perm()) | cpio.TypeDir,
+			}); err != nil {
+				return fmt.Errorf("could not write CPIO header: %w", err)
+			}
+
+			return nil
+		}
+
+		initrd.files = append(initrd.files, internal)
+
+		log.G(ctx).
+			WithField("file", path).
+			Trace("archiving")
+
+		reader, err := image.OpenPathFromSquash(path)
+		if err != nil {
+			return fmt.Errorf("could not open file: %w", err)
+		}
```

Comment:
Move this code block inside `:204`  to avoid it failing, that `if` block is the only place where this is used

Reply:
See PMs